sys/net/crosslayer/ng_pktbuf_static/ng_pktbuf_static.c

```diff
@@ -0,0 +1,430 @@
+        DEBUG("pktbuf: size (%u) == 0 || size == NG_PKTBUF_SIZE (%u)\n",
+              (unsigned)size, NG_PKTBUF_SIZE);
+        return NULL;
+    }
+    mutex_lock(&_mutex);
+    pkt = _create_snip(next, data, size, type);
+    mutex_unlock(&_mutex);
+    return pkt;
+}
+
+ng_pktsnip_t *ng_pktbuf_mark(ng_pktsnip_t *pkt, size_t size, ng_nettype_t type)
+{
+    ng_pktsnip_t *marked_snip;
+    /* size required for chunk */
+    size_t required_new_size = (size < sizeof(_unused_t)) ?
+                               _align(sizeof(_unused_t)) : _align(size);
+    mutex_lock(&_mutex);
+    if ((size == 0) || (pkt == NULL) || (size > pkt->size) || (pkt->data == NULL)) {
+        DEBUG("pktbuf: size == 0 (was %u) or pkt == NULL (was %p) or "
+              "size > pkt->size (was %u) or pkt->data == NULL (was %p)\n",
+              (unsigned)size, (void *)pkt, (unsigned)pkt->size, pkt->data);
+        mutex_unlock(&_mutex);
+        return NULL;
+    }
+    else if (size == pkt->size) {
+        pkt->type = type;
+        mutex_unlock(&_mutex);
+        return pkt;
+    }
+    marked_snip = _pktbuf_alloc(sizeof(ng_pktsnip_t));
+    if (marked_snip == NULL) {
+        DEBUG("pktbuf: could not reallocate marked section.\n");
+        mutex_unlock(&_mutex);
+        return NULL;
+    }
+    if ((size < required_new_size)) { /* would not fit unused marker => move data around */
+        void *new_data_marked, *new_data_rest;
+        new_data_marked = _pktbuf_alloc(size);
+        if (new_data_marked == NULL) {
+            DEBUG("pktbuf: could not reallocate marked section.\n");
+            _pktbuf_free(marked_snip, sizeof(ng_pktsnip_t));
+            mutex_unlock(&_mutex);
+            return NULL;
+        }
+        new_data_rest = _pktbuf_alloc(pkt->size - size);
+        if (new_data_rest == NULL) {
+            DEBUG("pktbuf: could not reallocate remaining section.\n");
+            _pktbuf_free(marked_snip, sizeof(ng_pktsnip_t));
+            _pktbuf_free(new_data_marked, size);
+            mutex_unlock(&_mutex);
+            return NULL;
+        }
+        memcpy(new_data_marked, pkt->data, size);
+        memcpy(new_data_rest, ((uint8_t *)pkt->data) + size, pkt->size - size);
+        _pktbuf_free(pkt->data, pkt->size);
+        marked_snip->data = new_data_marked;
+        pkt->data = new_data_rest;
+    }
+    else {
+        marked_snip->data = pkt->data;
+        pkt->data = ((uint8_t *)pkt->data) + size;
+    }
+    pkt->size -= size;
+    marked_snip->next = pkt->next;
+    marked_snip->size = size;
+    marked_snip->type = type;
+    marked_snip->users = 1;
+    pkt->next = marked_snip;
+    mutex_unlock(&_mutex);
+    return marked_snip;
+}
+
+int ng_pktbuf_realloc_data(ng_pktsnip_t *pkt, size_t size)
+{
+    size_t aligned_size = (size < sizeof(_unused_t)) ?
+                          _align(sizeof(_unused_t)) : _align(size);
+    mutex_lock(&_mutex);
+    assert((pkt == NULL) || (pkt->data == NULL) || _pktbuf_contains(pkt->data));
```

Comment:
This assert only checks if at least one condition is met, but should check that all of them are met.